File: src/test/scala/iterace/stage/TestThreadLocal.scala
package iterace.stage

import org.junit.Test
import iterace.IteRaceOption
import org.junit.Ignore

class TestThreadLocal extends RaceAbstractTest {

  val entryClass = "Lparticles/TestThreadLocal"
  override val options = {
    import iterace.IteRaceOption._
    Set[IteRaceOption](TwoThreads)
  }

  @Test def testNoRaceOnThreadLocalSet = expectNoRaces
  @Test def testNoRaceOnThreadLocalGet = expectNoRaces
  @Test def testNoRaceOnThreadLocalSetGet = expectNoRaces
  @Test def testNoRaceOnThreadLocalObject = expectNoRaces
  @Test def testNoRaceOnThreadLocalRetrivedObject = expectNoRaces
  @Test def testRaceOnThreadLocalRetrivedObject = expect("""
Loop: particles.TestThreadLocal.testRaceOnThreadLocalRetrivedObject(TestThreadLocal.java:87)

particles.Particle: particles.TestThreadLocal.testRaceOnThreadLocalRetrivedObject(TestThreadLocal.java:85)
 .x
   (a)  particles.TestThreadLocal$6.op(TestThreadLocal.java:92)
   (b)  particles.TestThreadLocal$6.op(TestThreadLocal.java:92)
""")
  
	@Test def testNoRaceOnObjectSetOutsideTheLoop = expectNoRaces
	@Test def testRaceOnThreadLocalReferredSharedObject = expect("""
Loop: particles.TestThreadLocal.testRaceOnThreadLocalReferredSharedObject(TestThreadLocal.java:116)

particles.Particle: particles.TestThreadLocal.testRaceOnThreadLocalReferredSharedObject(TestThreadLocal.java:114)
 .x
   (a)  particles.TestThreadLocal$8.op(TestThreadLocal.java:123)
   (b)  particles.TestThreadLocal$8.op(TestThreadLocal.java:123)
""")
//	@Test def  = expectNoRaces
//	@Test def  = expectNoRaces
//	@Test def  = expectNoRaces
}

File: src/subjects/particles/TestThreadLocal.java
package particles;

import extra166y.Ops;
import extra166y.ParallelArray;

public class TestThreadLocal {
	public ParallelArray<Particle> createParticleArray() {
		return ParallelArray.createUsingHandoff(new Particle[10],
				ParallelArray.defaultExecutor());
	}

	public void testNoRaceOnThreadLocalSet() {
		final ThreadLocal<Particle> shared = new ThreadLocal<Particle>();

		createParticleArray().replaceWithGeneratedValue(
				new Ops.Generator<Particle>() {
					@Override
					public Particle op() {
						shared.set(new Particle());
						return null;
					}
				});
	}

	public void testNoRaceOnThreadLocalGet() {
		final ThreadLocal<Particle> shared = new ThreadLocal<Particle>();
		shared.set(new Particle());

		createParticleArray().replaceWithGeneratedValue(
				new Ops.Generator<Particle>() {
					@Override
					public Particle op() {
						shared.get();
						return null;
					}
				});
	}

	public void testNoRaceOnThreadLocalSetGet() {
		final ThreadLocal<Particle> shared = new ThreadLocal<Particle>();
		shared.set(new Particle());

		createParticleArray().replaceWithGeneratedValue(
				new Ops.Generator<Particle>() {
					@Override
					public Particle op() {
						shared.set(new Particle());
						shared.get();
						return null;
					}
				});
	}

	public void testNoRaceOnThreadLocalObject() {
		final ThreadLocal<Particle> shared = new ThreadLocal<Particle>();

		createParticleArray().replaceWithGeneratedValue(
				new Ops.Generator<Particle>() {
					@Override
					public Particle op() {
						Particle particle = new Particle();
						shared.set(particle);
						particle.x = 7;
						return null;
					}
				});
	}
	
	public void testNoRaceOnThreadLocalRetrivedObject() {
		final ThreadLocal<Particle> shared = new ThreadLocal<Particle>();

		createParticleArray().replaceWithGeneratedValue(
				new Ops.Generator<Particle>() {
					@Override
					public Particle op() {
						shared.set(new Particle());
						shared.get().x = 7;
						return null;
					}
				});
	}
	
	public void testRaceOnThreadLocalRetrivedObject() {
		final ThreadLocal<Particle> shared = new ThreadLocal<Particle>();
		final Particle sharedParticle = new Particle();

		createParticleArray().replaceWithGeneratedValue(
				new Ops.Generator<Particle>() {
					@Override
					public Particle op() {
						shared.set(sharedParticle);
						shared.get().x = 7;
						return null;
					}
				});
	}
	
	public void testNoRaceOnObjectSetOutsideTheLoop() {
		final ThreadLocal<Particle> shared = new ThreadLocal<Particle>();
		shared.set(new Particle());

		createParticleArray().replaceWithGeneratedValue(
				new Ops.Generator<Particle>() {
					@Override
					public Particle op() {
						shared.get().x = 7;
						return null;
					}
				});
	}
	
	public void testRaceOnThreadLocalReferredSharedObject() {
		final ThreadLocal<Particle> shared = new ThreadLocal<Particle>();
		final Particle sharedParticle = new Particle();

		createParticleArray().replaceWithGeneratedValue(
				new Ops.Generator<Particle>() {
					@Override
					public Particle op() {
						Particle local = new Particle();
						local.origin = sharedParticle;
						shared.set(local);
						shared.get().origin.x = 7; // should race
						shared.get().x = 9; // should not race
						return null;
					}
				});
	}
}
